arangod/Aql/ClusterNodes.cpp

```diff
@@ -74,6 +77,662 @@ arangodb::velocypack::StringRef toString(GatherNode::SortMode mode) noexcept {
+      std::vector<SortRegister>& sortRegisters) noexcept
+    : _trx(trx),
+      _gatherBlockBuffer(gatherBlockBuffer),
+      _sortRegisters(sortRegisters) {
+  }
+
+  bool operator()(
+    std::pair<size_t, size_t> const& a,
+    std::pair<size_t, size_t> const& b
+  ) const;
+
+ private:
+  arangodb::transaction::Methods* _trx;
+  std::vector<std::deque<AqlItemBlock*>>& _gatherBlockBuffer;
+  std::vector<SortRegister>& _sortRegisters;
+}; // OurLessThan
+
+bool OurLessThan::operator()(
+    std::pair<size_t, size_t> const& a,
+    std::pair<size_t, size_t> const& b
+) const {
+  // nothing in the buffer is maximum!
+  if (_gatherBlockBuffer[a.first].empty()) {
+    return false;
+  }
+
+  if (_gatherBlockBuffer[b.first].empty()) {
+    return true;
+  }
+
+  TRI_ASSERT(!_gatherBlockBuffer[a.first].empty());
+  TRI_ASSERT(!_gatherBlockBuffer[b.first].empty());
+
+  for (auto const& reg : _sortRegisters) {
+    auto const& lhs = _gatherBlockBuffer[a.first].front()->getValueReference(a.second, reg.reg);
+    auto const& rhs = _gatherBlockBuffer[b.first].front()->getValueReference(b.second, reg.reg);
+    auto const& attributePath = reg.attributePath;
+
+    // Fast path if there is no attributePath:
+    int cmp;
+
+    if (attributePath.empty()) {
+#ifdef USE_IRESEARCH
+      TRI_ASSERT(reg.comparator);
+      cmp = (*reg.comparator)(reg.scorer.get(), _trx, lhs, rhs);
+#else
+      cmp = AqlValue::Compare(_trx, lhs, rhs, true);
+#endif
+    } else {
+      // Take attributePath into consideration:
+      bool mustDestroyA;
+      AqlValue aa = lhs.get(_trx, attributePath, mustDestroyA, false);
+      AqlValueGuard guardA(aa, mustDestroyA);
+      bool mustDestroyB;
+      AqlValue bb = rhs.get(_trx, attributePath, mustDestroyB, false);
+      AqlValueGuard guardB(bb, mustDestroyB);
+      cmp = AqlValue::Compare(_trx, aa, bb, true);
+    }
+
+    if (cmp == -1) {
+      return reg.asc;
+    } else if (cmp == 1) {
+      return !reg.asc;
+    }
+  }
+
+  return false;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// @class UnsortingGatherBlock
+/// @brief Execution block for gathers without order
+////////////////////////////////////////////////////////////////////////////////
+class UnsortingGatherBlock final : public ExecutionBlock {
+ public:
+  UnsortingGatherBlock(ExecutionEngine& engine, GatherNode const& en) noexcept
+    : ExecutionBlock(&engine, &en) {
+    TRI_ASSERT(en.elements().empty());
+  }
+
+  /// @brief shutdown: need our own method since our _buffer is different
+  int shutdown(int errorCode) override final {
+    DEBUG_BEGIN_BLOCK();
+    // don't call default shutdown method since it does the wrong thing to
+    // _gatherBlockBuffer
+    int ret = TRI_ERROR_NO_ERROR;
+    for (auto* dependency : _dependencies) {
+      int res = dependency->shutdown(errorCode);
+
+      if (res != TRI_ERROR_NO_ERROR) {
+        ret = res;
+      }
+    }
+
+    if (ret != TRI_ERROR_NO_ERROR) {
```

Comment:
`return ret;` would be the simplest solution here

Reply:
that was there before :-)
will fix

Reply:
done